Add Optimized Forward Kinematic Tree Implementation

Levi-Armstrong opened the pull request:
This is implemented base on the paper "A Forward Kinematics Data Structure for Efficient Evolutionary Inverse Kinematics"

mpowelson commented:
Just to help Google out a bit, I believe this is based on 

Starke, S., Hendrich, N., & Zhang, J. (2018). A Forward Kinematics Data Structure for Efficient Evolutionary Inverse Kinematics. In Computational Kinematics (pp. 560-568). Springer, Cham.

mpowelson commented on tesseract/tesseract_environment/test/tesseract_environment_unit.cpp:
```diff
@@ -65,8 +86,28 @@ void runContactManagerCloneTest(const tesseract_environment::Environment::Ptr& e
+template <typename S>
+void runAddandRemoveLinkTest()
 {
+  SceneGraph::Ptr scene_graph = getSceneGraph();
+  EXPECT_TRUE(scene_graph != nullptr);
+
+  auto env = std::make_shared<Environment>();
+  EXPECT_TRUE(env != nullptr);
+
+  bool success = env->init<S>(*scene_graph);
+  EXPECT_TRUE(success);
+
+  // Register contact manager
+  EXPECT_TRUE(env->registerDiscreteContactManager(tesseract_collision_bullet::BulletDiscreteBVHManager::name(),
+                                                  &tesseract_collision_bullet::BulletDiscreteBVHManager::create));
+  EXPECT_TRUE(env->registerContinuousContactManager(tesseract_collision_bullet::BulletCastBVHManager::name(),
+                                                    &tesseract_collision_bullet::BulletCastBVHManager::create));
+
+  // Set Active contact manager
+  EXPECT_TRUE(env->setActiveDiscreteContactManager(tesseract_collision_bullet::BulletDiscreteBVHManager::name()));
+  EXPECT_TRUE(env->setActiveContinuousContactManager(tesseract_collision_bullet::BulletCastBVHManager::name()));
```
This code is repeated a few times. Consider putting in a utility

Levi-Armstrong replied:
Done.

mpowelson commented on tesseract/tesseract_environment/test/tesseract_environment_unit.cpp:
```diff
@@ -324,172 +530,52 @@ void runCloneTest(const tesseract_environment::Environment::ConstPtr& env)
-  EXPECT_TRUE(success);
-
-  runCloneTest(env);
+  runCloneTest<KDLStateSolver>();
```
Should this be changed to EnvClone to match?

Levi-Armstrong replied:
done.

mpowelson commented on tesseract/tesseract_environment/src/ofkt/ofkt_state_solver.cpp:
```diff
@@ -0,0 +1,805 @@
+  template <class u, class g>
+  void discover_vertex(u vertex, g graph)
+  {
+    // Get incomming edges
```
spelling `incoming`

Levi-Armstrong replied:
fixed.

mpowelson commented on tesseract/tesseract_environment/src/ofkt/ofkt_state_solver.cpp:
```diff
@@ -0,0 +1,805 @@
+      nodes_[joint_name] = std::move(n);
+      joint_names_.push_back(joint_name);
+      limits.push_back(
+          std::make_pair(joint_name, std::array<double, 2>({ joint->limits->lower, joint->limits->upper })));
```
Will this work or is some limit assumed? Should a warning be printed?

Levi-Armstrong replied:
There should always be limits but I have added a check in side the revolute constructor.

mpowelson commented on tesseract/tesseract_environment/include/tesseract_environment/ofkt/ofkt_node.h:
```diff
@@ -0,0 +1,79 @@
+
+namespace tesseract_environment
+{
+class OFKTNode
```
class doxygen

Levi-Armstrong replied:
Done

mpowelson commented on tesseract/tesseract_environment/include/tesseract_environment/ofkt/ofkt_state_solver.h:
```diff
@@ -0,0 +1,132 @@
+
+namespace tesseract_environment
+{
+class OFKTStateSolver : public StateSolver
```
class doxygen

Levi-Armstrong replied:
Done.

mpowelson commented on tesseract/tesseract_environment/test/tesseract_environment_unit.cpp:
```diff
@@ -10,6 +10,7 @@ TESSERACT_COMMON_IGNORE_WARNINGS_POP
 
 #include <tesseract_environment/core/types.h>
 #include <tesseract_environment/kdl/kdl_state_solver.h>
+#include <tesseract_environment/ofkt/ofkt_state_solver.h>
```
Outside the scope of this PR, but it might be interesting to make some benchmarks for the state solvers like we have for collision

Levi-Armstrong commented on tesseract/tesseract_environment/include/tesseract_environment/ofkt/ofkt_node.h:
```diff
@@ -0,0 +1,79 @@
+#ifndef TESSERACT_ENVIRONMENTOFKT_NODE_H
```
updated

Levi-Armstrong commented on tesseract/tesseract_environment/include/tesseract_environment/ofkt/ofkt_state_solver.h:
```diff
@@ -0,0 +1,132 @@
+#ifndef TESSERACT_ENVIRONMENT_OFKT_STATE_SOLVER_H
```
updated

Levi-Armstrong commented on tesseract/tesseract_environment/include/tesseract_environment/ofkt/ofkt_node.h:
```diff
@@ -0,0 +1,79 @@
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+  using UPtr = std::unique_ptr<OFKTNode>;
+
```
done

Levi-Armstrong commented on tesseract/tesseract_environment/src/ofkt/ofkt_state_solver.cpp:
```diff
@@ -0,0 +1,805 @@
+
+namespace tesseract_environment
+{
+class OFKTBaseNode : public OFKTNode
```
Done.

Levi-Armstrong commented:
@mpowelson I believe it is ready to test if you want to. All of the unit tests pass and I have added some additional test to test other aspects of the state solver to make sure the EnvState contains the correct information when modifying the environment. Also added a forward kinematics tests, but not everything is currently being tested related to all possible environment commands.

mpowelson commented:
@Levi-Armstrong I will give this a test. For what it is worth, we will need to add a way to switch between them in a Tesseract. For now I will just hardcode it while I test. Where you imagining this would become the default?

mpowelson commented on tesseract/tesseract_environment/src/ofkt/ofkt_state_solver.cpp:
```diff
@@ -0,0 +1,591 @@
+{
+  for (std::size_t i = 0; i < joint_names.size(); ++i)
+  {
+    nodes_[joint_names[i]]->storeJointValue(joint_values(static_cast<long>(i)));
```
@Levi-Armstrong I'm getting a segfault here. joint_names[i] is not in the map, so it creates one. Then ->storeJointValue is derefencing a nullptr.

mpowelson replied:
While this should probably be fixed. It looks like the cause was that clone does not populate nodes_

mpowelson replied:
It looks like cloneHelper is not being called.

mpowelson commented:
@Levi-Armstrong I'm getting a lot of warnings like this. Any idea why? Some sort of roundoff somewhere?

`Joint value (3.143513) is above the upper limits (3.142593), setting to upper limit.`

mpowelson commented:
The good news is that addLink now appears to take 4ms instead of 700ms!

Levi-Armstrong commented:
> @Levi-Armstrong I'm getting a lot of warnings like this. Any idea why? Some sort of roundoff somewhere?
> 
> `Joint value (3.143513) is above the upper limits (3.142593), setting to upper limit.`

So I added a check in the revolute joint node to make sure you do not set the joint outside its limits. Is the number printed out the correct limit? Also what planner are you using?

mpowelson commented:
Interpolator, collision check, trajopt, collision check

The check makes sense, but the numbers are just barely outside the limits. Is this happening because of the resolution of trajopt? If so, perhaps we should make this a debug message.

Levi-Armstrong commented:
I think this may have exposed some issue with  the planners. I would have expected it to be closer. Are they all close like this?

mpowelson commented:
Some of them are close enough that the 6 digits that are printing look the same. I haven't seen any large differences

mpowelson commented:
Regardless, nice job. This works way better than the KDL solver.

How should we switch between them in a tesseract? We could template the init like we do with the environment. Or we could have an enum switch or something.

Levi-Armstrong commented:
Lets switch it to debug message. Also we could have it print a warning if it is outside by some tolerance, maybe the float epsilon?

Levi-Armstrong commented:
I was thinking we would set this as the default, but it would be good to have a way to switch.

mpowelson commented:
I tried the epsilon tolerance, but it is outside by more than that.

I'm fine with setting it as default. Then it will get well tested while we work on feature/CommandLanguage
